horaedb/metric_engine/src/manifest.rs

```diff
@@ -184,6 +203,264 @@ impl Manifest {
+
+impl Into<SstFile> for SnapshotRecordV1 {
+    fn into(self) -> SstFile {
+        let file_meta = FileMeta {
+            max_sequence: self.id(),
+            num_rows: self.num_rows,
+            size: self.size,
+            time_range: TimeRange::new(
+                Timestamp::from(self.time_range_start),
+                Timestamp::from(self.time_range_end),
+            ),
+        };
+        SstFile::new(self.id(), file_meta)
+    }
+}
+
+struct Snapshot {
+    header: SnapshotHeader,
+    inner: Bytes,
+}
+
+impl Default for Snapshot {
+    // create an empty Snapshot
+    fn default() -> Self {
+        let header = SnapshotHeader::new(0);
+        Self {
+            header,
+            inner: Bytes::new(),
+        }
+    }
+}
+
+impl Snapshot {
+    pub fn try_from_bytes(bytes: Bytes) -> Result<Self> {
+        if bytes.is_empty() {
+            return Ok(Snapshot::default());
+        }
+        let header = SnapshotHeader::try_from_bytes(bytes.as_bytes())?;
+        let header_length = header.length as usize;
+        if header_length > 0
+            && (header_length % SnapshotRecordV1::LENGTH != 0
+                || header_length + SnapshotHeader::LENGTH != bytes.len())
+        {
+            return Err(anyhow!(
+                "create snapshot from bytes failed, invalid bytes, header length = {}, total length: {}",
+                header_length,
+                bytes.len()
+            ).into());
+        }
+        Ok(Self {
+            header,
+            inner: bytes,
+        })
+    }
+
+    pub fn to_sstfiles(&self) -> Result<Vec<SstFile>> {
+        if self.header.length == 0 {
+            return Ok(Vec::new());
+        } else {
+            let buf = self.inner.as_bytes();
+            let mut result: Vec<SstFile> = Vec::new();
+            let mut index = SnapshotHeader::LENGTH;
+            while index < buf.len() {
+                let record =
+                    SnapshotRecordV1::try_from(&buf[index..index + SnapshotRecordV1::LENGTH])?;
+                index += SnapshotRecordV1::LENGTH;
+                result.push(record.into());
+            }
+
+            Ok(result)
+        }
+    }
+
+    pub fn dedup_sstfiles(&self, sstfiles: &mut Vec<SstFile>) -> Result<()> {
+        let buf = self.inner.as_bytes();
+        let mut ids = HashSet::new();
+        let mut index = SnapshotHeader::LENGTH;
+        while index < buf.len() {
+            let record = SnapshotRecordV1::try_from(&buf[index..index + SnapshotRecordV1::LENGTH])?;
+            index += SnapshotRecordV1::LENGTH;
+            ids.insert(record.id());
+        }
+        sstfiles.retain(|item| !ids.contains(&item.id()));
+
+        Ok(())
+    }
+
+    pub fn merge_sstfiles(&mut self, sstfiles: Vec<SstFile>) {
+        // update header
+        self.header.length += (sstfiles.len() * SnapshotRecordV1::LENGTH) as u64;
+        // final snapshot
+        let mut snapshot = vec![1u8; SnapshotHeader::LENGTH + self.header.length as usize];
```

Comment:
Why to init all bytes to 1?

Reply:
It should be 0u8. It's simply a mistake.